### PR TITLE
Serialize common Django objects

### LINKED ISSUE #2808: Serialize Django QuerySet for error responses
> > Trying to delete a schema mapping field results in this error:
> > ```
> > Object of type QuerySet is not JSON serializable
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > image
> > _Just an observation. Not tremendously helpful or informative, just reporting._
> 
> There are some cases where something in the original error is not Json Serializable and thus can not be directly sent as a Json response. To get around this, I manually catch the error that is thrown when trying to serialize the object and serialize that instead to json and sent it to the frontend. Based on the traceback, I still know you were trying to delete something.
> 
> I can try and include more information for these errors specifically if we need, but regardless the information we send will probably be limited.
> 
> Actually, I can write up our own way of serializing specifically django's QuerySet. We can not anticipate every object, but QuerySet might appear often so it may be worth the effort

- https://github.com/specify/specify7/pull/2702#issuecomment-1379720672

Currently, when the backend encounters something in an Exception that can not be serialized to json, it catches the thrown TypeError and sends that instead. While this is still better than a `500 error`,  it is far from helpful to any user who does not know how to read the traceback. 

While we can not anticipate every object that is not json serializable, having a representation for Django's representation of a query ([QuerySet](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#django.db.models.query.QuerySet)), would be helpful.

### PR DESCRIPTION
**Pull request by melton-jason**
Fixes #2808 

These changes serialize two common Django Objects, Querysets and Models. For Querysets, the data returned is a list of dictionaries which contain the keys `Model`, `Id`, and `Non-Null Fields`. 
For single instance Model objects, a single dictionary is returned containing _every_ fields value for that object. 

In addition, the backed should try and convert the data of every exception into a string before parsing it as JSON. This should hopefully eliminate most error messages which say `Object of type <object> is not Json serializable`.

For testing, reference #1385, #2511, #2886, #2884

### CODE REVIEW

File: specifyweb/middleware/general.py [hunk 1]
```diff
@@ -51,3 +59,39 @@ def process_exception(self, request, exception) -> Optional[http.HttpResponse]:
             except TypeError as e:
                 exception = SpecifyExceptionWrapper(e)
                 return http.HttpResponse(exception.to_json(), status=exception.status_code)
+
+
+def serialize_django_obj(django_obj: models.QuerySet or models.Model) -> List[Dict[str, Any]] or Dict[str, Any]:
```
Comment (maxxxxxdlp) [inline specifyweb/middleware/general.py:64]: There is a function in `./specify/api.py` called `_obj_to_data`. Can you check if it is related to what you are doing here? If so, would it make sense to merge that function with this one?

Comment (melton-jason) [reply specifyweb/middleware/general.py:64]: Yes I can merge the two. Currently `_obj_to_data` does not handle serializing QuerySets, but I can always repurpose the logic. 

The main concern I have with this is that `_obj_to_data` returns every field of the associated object. If the object has any relationships, it includes every field in the associated table (`_obj_to_data` does the heavy work for creating the responses for the main specify endpoint)

For example, here is a (nonformatted json) agent with 2 addresses:

serialized_agent

There is the argument that the data is important, but I feel that (more often than not) most of the data from the function does not need to be viewed and it _does_ clutter a _lot_ of the space and can hinder viewing _all_ of the records. The most important information will be 'lost' in the clutter

Comment (melton-jason) [reply specifyweb/middleware/general.py:64]: The question is should we truncate the data to the most 'important' parts and make the data more readable, or show _all_ of the related data but make everything harder to find/identify?

**Comment by maxxxxxdlp** [comment]
We have the ability to display short "formatted" records (it's the `format` button on the front-end)
Back-end also has that capability (don't know where in the code it is but it is used in the query builder)

**Comment by melton-jason** [comment]
> How can I test this?

@grantfitzsimmons 
The changes here are two-fold. 
There was only one (to my knowledge) BusinessRule which was affected by these changes: uniqueness rules.

For other cases, you may need to recreate a known issue and view the response. Look for any issues that raise common django exceptions such as Integrity Errors, Protected Errors, Programming Errors, Validation Errors, Field Errors, etc.
Furthermore, these changes should attempt and serialize _any_ backend exception by parsing it into a string (So for instance, any SQL Alchemy errors would be serialized as well), but the most common are django ones.

Here are some known issues regarding Uniqueness Rules:
- https://github.com/specify/specify7/issues/2884
- https://github.com/specify/specify7/issues/2859

Here are some known issues regarding django errors:
- https://github.com/specify/specify7/issues/2886
- https://github.com/specify/specify7/issues/1385
- https://github.com/specify/specify7/issues/2511

I am sure there are others I am missing.
You can reproduce the errors on this branch and see the output.

**Comment by grantfitzsimmons** [comment]
**On this branch:**

#1385:

image

https://coldfish-issue-2808.test.specifysystems.org/specify/query/37/

**Comment by grantfitzsimmons** [comment]
**On this branch:**

#2884 

image

_Definitely a big improvement!_

**Comment by grantfitzsimmons** [comment]
**On this branch:**

It looks the same as it does on the before branch.

#2511

https://herbrbge11723-issue-2808.test.specifysystems.org/specify/view/borrow/1364/

image


[Specify 7 Crash Report - 2023-02-13T15 00 53.813Z.txt](https://github.com/specify/specify7/files/10723418/Specify.7.Crash.Report.-.2023-02-13T15.00.53.813Z.txt)

**Comment by grantfitzsimmons** [comment]
**On this branch:**

image

https://herbrbge11723-issue-2808.test.specifysystems.org/specify/view/collectionobject/new/

Create a preparation with a barcode that = `E01004897`

**Comment by grantfitzsimmons** [comment]
For the record, we should not close any issue outside of #2884 when this is merged

**Review by grantfitzsimmons — APPROVED** [review]
I think it looks good. Some improvements, some remain about the same.

**Comment by grantfitzsimmons** [comment]
@melton-jason Please review my replies and merge if you feel comfortable!

**Comment by melton-jason** [comment]
> **On this branch:**
> 
> It looks the same as it does on the before branch.
> 
> #2511
> 
> https://herbrbge11723-issue-2808.test.specifysystems.org/specify/view/borrow/1364/
> 
> image
> 
> [Specify 7 Crash Report - 2023-02-13T15 00 53.813Z.txt](https://github.com/specify/specify7/files/10723418/Specify.7.Crash.Report.-.2023-02-13T15.00.53.813Z.txt)

Fixed. QuerySets should now properly display some pertinent information about their contained objects

**Comment by grantfitzsimmons** [comment]
> > **On this branch:**
> > It looks the same as it does on the before branch.
> > #2511
> > https://herbrbge11723-issue-2808.test.specifysystems.org/specify/view/borrow/1364/
> > image
> > [Specify 7 Crash Report - 2023-02-13T15 00 53.813Z.txt](https://github.com/specify/specify7/files/10723418/Specify.7.Crash.Report.-.2023-02-13T15.00.53.813Z.txt)
> 
> Fixed. QuerySets should now properly display some pertinent information about their contained objects

image

[Specify 7 Crash Report - 2023-02-24T15 15 39.496Z.txt](https://github.com/specify/specify7/files/10825664/Specify.7.Crash.Report.-.2023-02-24T15.15.39.496Z.txt)

Super interesting response this time:
```
ProtectedError at /api/specify/borrow/1364/ ("Cannot delete some instances of model 'Addressofrecord' because they are referenced through protected foreign keys: 'Borrow.addressofrecord'.", {, , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , }) Request Method: DELETE Request URL: http://herbrbge11723-issue-2808.test.specifysystems.org/api/specify/borrow/1364/ Django Version: 3.2.15 Python Executable: /opt/specify7/ve/bin/python3.8 Python Version: 3.8.0 Python Path: ['/opt/specify7', '/opt/specify7', '/opt/specify7/ve/bin', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '/opt/specify7/ve/lib/python3.8/site-packages', '/opt/specify7'] Server time: Fri, 24 Feb 2023 15:15:14 -0600 Installed Applications: ('django.contrib.sessions', 'django.contrib.staticfiles', 'django.contrib.contenttypes', 'django.contrib.auth', 'specifyweb.specify', 'specifyweb.permissions', 'specifyweb.accounts', 'specifyweb.stored_queries', 'specifyweb.businessrules', 'specifyweb.express_search', 'specifyweb.context', 'specifyweb.attachment_gw', 'specifyweb.frontend', 'specifyweb.barvis', 'specifyweb.report_runner', 'specifyweb.interactions', 'specifyweb.workbench', 'specifyweb.notifications', 'specifyweb.export', 'specifyweb.raven_placeholder') Installed Middleware: ['django.middleware.gzip.GZipMiddleware', 'django.contrib.sessions.middleware.SessionMiddleware', 'django.middleware.locale.LocaleMiddleware', 'django.middleware.common.CommonMiddleware', 'django.middleware.csrf.CsrfViewMiddleware', 'django.contrib.auth.middleware.AuthenticationMiddleware', 'specifyweb.context.middleware.ContextMiddleware', 'specifyweb.permissions.middleware.PermissionsMiddleware', 
```

**Comment by melton-jason** [comment]
> > > **On this branch:**
> > > It looks the same as it does on the before branch.
> > > #2511
> > > https://herbrbge11723-issue-2808.test.specifysystems.org/specify/view/borrow/1364/
> > > image
> > > [Specify 7 Crash Report - 2023-02-13T15 00 53.813Z.txt](https://github.com/specify/specify7/files/10723418/Specify.7.Crash.Report.-.2023-02-13T15.00.53.813Z.txt)
> > 
> > 
> > Fixed. QuerySets should now properly display some pertinent information about their contained objects
> 
> image
> 
> [Specify 7 Crash Report - 2023-02-24T15 15 39.496Z.txt](https://github.com/specify/specify7/files/10825664/Specify.7.Crash.Report.-.2023-02-24T15.15.39.496Z.txt)
> 
> Super interesting response this time:
> 
> ```
> ProtectedError at /api/specify/borrow/1364/ ("Cannot delete some instances of model 'Addressofrecord' because they are referenced through protected foreign keys: 'Borrow.addressofrecord'.", {, , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , }) Request Method: DELETE Request URL: http://herbrbge11723-issue-2808.test.specifysystems.org/api/specify/borrow/1364/ Django Version: 3.2.15 Python Executable: /opt/specify7/ve/bin/python3.8 Python Version: 3.8.0 Python Path: ['/opt/specify7', '/opt/specify7', '/opt/specify7/ve/bin', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '/opt/specify7/ve/lib/python3.8/site-packages', '/opt/specify7'] Server time: Fri, 24 Feb 2023 15:15:14 -0600 Installed Applications: ('django.contrib.sessions', 'django.contrib.staticfiles', 'django.contrib.contenttypes', 'django.contrib.auth', 'specifyweb.specify', 'specifyweb.permissions', 'specifyweb.accounts', 'specifyweb.stored_queries', 'specifyweb.businessrules', 'specifyweb.express_search', 'specifyweb.context', 'specifyweb.attachment_gw', 'specifyweb.frontend', 'specifyweb.barvis', 'specifyweb.report_runner', 'specifyweb.interactions', 'specifyweb.workbench', 'specifyweb.notifications', 'specifyweb.export', 'specifyweb.raven_placeholder') Installed Middleware: ['django.middleware.gzip.GZipMiddleware', 'django.contrib.sessions.middleware.SessionMiddleware', 'django.middleware.locale.LocaleMiddleware', 'django.middleware.common.CommonMiddleware', 'django.middleware.csrf.CsrfViewMiddleware', 'django.contrib.auth.middleware.AuthenticationMiddleware', 'specifyweb.context.middleware.ContextMiddleware', 'specifyweb.permissions.middleware.PermissionsMiddleware', 
> ```
> 
> https://herbrbge11723-issue-2808.test.specifysystems.org/specify/view/borrow/1364/

That output screen is a known bug in production when Debug mode is enabled. This will have to be tested with Debug mode disabled until that bug is fixed and the Django error page can be displayed once again. 
More specifically, we need to make sure the output is not `Borrow.addressofrecord'.", {, , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , })`.

**Comment by maxxxxxdlp** [comment]
@melton-jason since this is approved, is it ready for merging?

**Comment by melton-jason** [comment]
I will quickly make sure the output is correct when debug mode is disabled. If so, then I will merge this PR

**Comment by maxxxxxdlp** [comment]
If it's working, I can resolve the merge commit and merge it

**Comment by melton-jason** [comment]
The output is correct. 
serialized_django_obj

It will show this output in the Django dialog once we get that bug fixed!

**Pull request merged**